packages/nuxt/src/core/app.ts

```diff
@@ -125,20 +125,31 @@ async function resolveApp (nuxt: Nuxt, app: NuxtApp) {
     }))
   }
 
-  // Resolve plugins
-  app.plugins = [
-    ...nuxt.options.plugins.map(normalizePlugin)
-  ]
+  // Re-initialise plugins
+  app.plugins = nuxt.options.plugins.map(normalizePlugin)
+
+  // Track scanned plugins separately so we can apply a sorting mechanism to them
+  const scannedPlugins: [string, NuxtPlugin][] = []
   for (const config of nuxt.options._layers.map(layer => layer.config)) {
-    app.plugins.push(...[
-      ...(config.plugins || []),
-      ...config.srcDir
-        ? await resolveFiles(config.srcDir, [
-          `${config.dir?.plugins || 'plugins'}/*.{ts,js,mjs,cjs,mts,cts}`,
-          `${config.dir?.plugins || 'plugins'}/*/index.*{ts,js,mjs,cjs,mts,cts}` // TODO: remove, only scan top-level plugins #18418
-        ])
-        : []
-    ].map(plugin => normalizePlugin(plugin as NuxtPlugin)))
+    for (const plugin of config.plugins || []) {
+      app.plugins.push(normalizePlugin(plugin as NuxtPlugin | string))
+    }
+
+    if (config.srcDir) {
+      const pluginDir = join(config.srcDir, config.dir?.plugins || 'plugins')
+      for (const file of await resolveFiles(pluginDir, [
+        '*.{ts,js,mjs,cjs,mts,cts}',
+        '*/index.*{ts,js,mjs,cjs,mts,cts}' // TODO: remove, only scan top-level plugins #18418
+      ])) {
+        scannedPlugins.push([file.replace(pluginDir, ''), normalizePlugin(file)])
+      }
+    }
+  }
+
+  // Sort scanned plugins by their names so a scanned `01.plugin.ts` will always run before finished `02.plugin.ts`
+  // and append to the plugin array
+  for (const [_name, plugin] of scannedPlugins.sort(([a], [b]) => a.localeCompare(b))) {
```

Comment:
As per https://github.com/nuxt/nuxt/pull/22889#pullrequestreview-1600781599, shouldn't this be _within_ the for loop to sort each layer by itself?

Sorting all plugins might sound reasonable when authoring all layers but when extending with 3rd party layers, this order can be easily messed up with numbering. I think we shall at least have an opt-in (or opt-out if you see benefits of this as default) for layers meta config. (also the fact that, plugins shall not depend on ordering as much as possible and prefer hooks...)

Reply:
thanks - I appreciate the potential breaking nature of this pr so really appreciate feedback on it.

I think the main scenario is that if a user creates a plugin, they would expect to be able to access any injections from layers, and currently all layer plugins are loaded _after_ all user plugins. a lighter weight solution might be to simply reverse that order and load layer plugins before user plugins? I think this might also be good enough.

escape hatch for layer authors is mostly that plugins can be specified in `nuxt.config` and they will then be loaded before any scanned plugins.

Reply:
Reversing order of layers seems a good idea! Because indeed by nature layers are supposed to be overridable and accessible by higher order